### DISCUSSION TITLE
Disabling SSR for a Component

### DISCUSSION THREAD

**Discussion by ashikmeerankutty** [remix-run/remix#1023] [Q&A] [answered]
## Problem

I am trying to use a line chart component from `@ant-design/charts`. I am getting an error `ReferenceError: navigator is not defined `. When I search for the issue I was able to find that navigator is not available in SSR. So I need to disable SSR for this component. I tried using the `ClientOnly` #180 . But It didn't solve my issue. 

## Solutions I have tried

I also tried conditionally rendering the component by checking `(typeof window !== "undefined") && window.navigator` that also didn't fix the issue. Then I found that next js solves this issue by using [dynamic imports](https://github.com/JedWatson/react-codemirror/issues/77#issuecomment-645362423). I couldn't find a similar API in remix.

**Comment by kentcdodds** [comment, marked as answer]
You can put the import for the component in a `app/components/charts.client.tsx` and that will prevent that code from being used in the server.

**Reply by ashikmeerankutty** [reply]
Tried that, Its throwing an un expected error: 

```
Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.
```

Although I am importing component properly. 

Here is a repo where the issue is reproducible: https://github.com/ashikmeerankutty/client-test

`/chart` route throws the error.

**Reply by kentcdodds** [reply]
Here you go: https://github.com/ashikmeerankutty/client-test/pull/1

**Reply by ashikmeerankutty** [reply]
Thank you so much @kentcdodds , that works.
So, for client side rendering we have to use both `x.client.tsx` with `ClientOnly`.

But why does the the element at `chart.client.tsx` throws the error `Element type is invalid`.

**Reply by kentcdodds** [reply]
It's because on the server when there's a client import all those values are undefined so on the server you were rendering undefined. You need to not render it on the server which is where the ClientOnly component or the userEffect comes in handy

**Reply by ashikmeerankutty** [reply]
Thanks, Got it.

**Reply by expressodaily** [reply]
Great! thanks

**Comment by sidsarasvati** [comment]
Wow! Glad I found this answer. Almost moved to nextjs framework as I got frustrated trying to make p5 library work with remix run. 
Appreciate remix_run team keeping the discussion here well organized and maintained.
